proposed/2021/BulkRestoreCoordination.md

```diff
@@ -0,0 +1,165 @@
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+#### Target scenario
+
+Ex:
+Given a solution with 3 projects.
+Proj A -> Proj B -> Proj C.
+
+Say a Directory.Build.Props is edited that affect *all* projects.
+Say the nominations come in the following order:
+
+Proj B
+Proj C
+Proj A
+
+- When the nomination for B comes in, NuGet will trigger a restore that will affect projects B & A, given that A depends on B.
+- While that first restore runs, it possible that C & A nominations have arrived. When NuGet receives these nominations and the previous restore completes, which happens last, a new restore will be kicked off, with the nomination data for C & A.
+- This second restore will affect A, B & C. Project A & B have been restored twice, thus causing unnecessary work.
+- In the ideal scenario, only 1 restore for each 3 projects is done.
+
+#### Restore batching
+
+At solution load, project-system runs DT builds and call the `IVsSolutionRestoreService` to nominate a project.
+NuGet is aware of *all* projects that are supposed to nominate, so NuGet can wait for *all* projects to be nominated.
+Currently NuGet has a sliding window to account for potentially failed projects.
+
+When the solution is loaded, NuGet doesn't have an additional heuristic to determine whether there are other projects that might need a restore done.
+The proposal is that each time NuGet determines a restore might need run, it checks whether there are *any* projects with pending *design time builds*, that *might* nominate. If yes, NuGet will wait.
+
+#### New interfaces
+
+##### IVsSolutionRestoreService4
+
+```cs
+    /// <summary>
+    /// Represents a package restore service API for integration with a project system.
+    /// Implemented by NuGet.
+    /// </summary>
+    [ComImport]
+    [Guid("72327117-6552-4685-BD7E-9C40A04B6EE5")]
+    public interface IVsSolutionRestoreService4
+    {
+        /// <summary>
+        /// A project system can call this service (optionally) to register itself to coordinate restore.
+        /// Each project can only register once. NuGet will call into the source to wait for nominations for restore.
+        /// NuGet will remove the registered object when a project is unloaded.
+        /// </summary>
+        /// <param name="restoreInfoSource">Represents a project specific info source</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        Task RegisterRestoreInfoSourceAsync(IVsProjectRestoreInfoSource restoreInfoSource, CancellationToken cancellationToken);
+    }
+```
+
+##### IVsProjectRestoreInfoSource
+
+```cs
+    /// <summary>
+    /// Represents a service that NuGet can use to ask a project-system whether there are any pending design time builds.
+    /// Implemented by the project-system.
+    /// </summary>
+    [ComImport]
+    [Guid("35AD5FF2-6AB7-48E9-BCC0-189042410FA6")]

```

Comment:
Did you generate this random new Guid?

Reply:
Yep, we generate a new guid to associate with that interface anytime we add a new interop interface. 

If you take a look at https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop, all interfaces have a GUID.